### ISSUE TITLE
Nominate restore should get PackageReferences from a target instead of evaluation.

### ISSUE THREAD

**Issue by srivatsn** [dotnet/roslyn-project-system#653]
Currently the set of PackageReference items that we send to NuGet to restore are the ones that are from the ProjectRuleSource - meaning they are from a static MSBuild evaluation. Instead we should have a Target called ResolvePackageReferences or some such that we can call that could be overridden to dynamic provide PackageRefs to be restored. This will enable other package managers like Paket to participate in background auto-restore.

**Comment by forki** [comment]
Thanks for the hard work you guys do to make it an open platform.  In Paket 4 alpha we currently do the following:

    <Target Name="PaketRestore" BeforeTargets="_GenerateRestoreGraphWalkPerFramework" >
      <Exec Command="$(MSBuildThisFileDirectory)\paket.exe restore --project $(MSBuildProjectFullPath)" />

      
      <PaketRestoreTask
        ProjectUniqueName="$(MSBuildProjectFullPath)"
        PackageReferences="@(PackageReference)"
        TargetFrameworks="$(TargetFramework)">
        <Output TaskParameter="NewPackageReferences" ItemName="PackageReference" />
        <Output TaskParameter="AlternativeConfigFile" ItemName="RestoreConfigFile" />        
      </PaketRestoreTask>
    </Target>

I think `_GenerateRestoreGraphWalkPerFramework` comes from `NuGet.targets`. Since it's marked with underscore I assume we should not take that one ;-) 

Anyways, our current approach is to add PackageReference entries in memory and to override RestoreConfigFile (which I believe is the path to NuGet.Config).

**Comment by davidfowl** [comment]
@srivatsn This would be nice. Would the IDE still work well or would we lose any efficiency because we need to re-run the target on any dependency change? How will that work actually? 

Lets say I have an external that has dependencies and I override the right targets so that msbuild can determine the right package reference nodes. How do I communicate to the project system that when I change my external file it should re-evaluate the package refs?

**Comment by forki** [comment]
How would that work today? Something like a file watcher on the csproj? Could we register more files to watch? Can we just delete obj folder or wherever design builds are caching?

**Comment by srivatsn** [comment]
Yes we have a [file watcher](https://github.com/dotnet/roslyn-project-system/blob/b22cf2e196665e747a00c60d53aba0865567f329/src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/ProjectLockFileWatcher.cs) that monitors the nuget lock file and tells CPS that it needs to reevaluate. Since your PaketRestore will trigger a NuGet restore, you should be fine I think.

Today, without running any targets, reading the items from the static evaluation is faster than if we did a design-time build to get the packagereferences (i.e if we call the target to resolve). However, given this is a background operation, I'm not sure how much of an experience difference there will be - we'll have to implement and see. One approach could be that we pass NuGet the data from evaluation first and when the design-time results come in, we pass in additional references but that's extra state tracking that would be nice to avoid if the perf was acceptable.

**Comment by forki** [comment]
I *believe* (please correct me when I'm wrong) @davidfowl meant a watcher on project.json itself, right? So that a save on the project file would trigger restore in the background. That's something paket would not do since we are editing paket.references files.

**Comment by srivatsn** [comment]
Ah yes. In the case of csproj, CPS watches the project file and causes reevaluations to happen. Changes to other files don't cause reevaluation. Today, you'll need to write a file watcher like the one mentioned above. However, we should probably figure out some scheme for marking msbuild items in a way that changing their contents trigger a reevaluation. Tagging @lifengl as well as this would be a CPS feature.

**Comment by srivatsn** [comment]
We now get the packagereferences from a target called `CollectProjectReferences`. Overriding that target should be sufficient to restore both in the commandline and in VS.

**Comment by davkean** [comment]
Reopening this - I don't understand how this is supposed to work. `CollectProjectReferences` isn't called by the command-line restore. VS restore and command-line restore should call the same targets to gather package references.

**Comment by srivatsn** [comment]
It used to be run in the commandline restore case but looks like a refactoring in the SDK broke that. I think the right place to put this requirement is in the NuGet.targets before it computes the RestoreStyle [here](https://github.com/NuGet/NuGet.Client/blob/68245312b050c5b992120448353eba3e20cdc414/src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets#L344) so that if a target injected some PackageReferences, NuGet will know to choose the PackageReference style over others. @emgarten do you agree?

The change I'm proposing is to make this line be
`DependsOnTargets="_GetProjectJsonPath;CollectPackageReferences"`

**Comment by emgarten** [comment]
Is there a way to ensure that ``CollectPackageReferences`` doesn't return ``PackageReference`` items from props/targets in restored packages? Bringing these in, or any props/targets from a package at restore time means that the restore result is different between the first and second restores, and potentially beyond that if new packages keep showing up.

NuGet currently excludes these by setting ``ExcludeRestorePackageImports=true`` from the command line. The project system should do something similar.

Would this target be available for legacy CSProj or would restore need to fall back to reading them directly if this is not a NETCore SDK project?

I opened https://github.com/NuGet/Home/issues/4967 to track this for command line restore.

**Comment by srivatsn** [comment]
> But I would expect the project system to start using this also to stay consistent with command line restore.

I don't understand this. Are you saying that when we do background restore in VS, we need to set this property somewhere? What does it mean for the project system to use this property?

**Comment by davkean** [comment]
@srivatsn It means, we should make sure ExcludeRestorePackageImports=true when we call CollectPackageReferences to be consistent with command-line. That way we don't accidentally gather package references from targets brought in by packages.

This actually leads to a different concern if this; CPS groups design-time build targets _together_ at the same time; we wouldn't want _CompileDesignTime_ to be run at the same as this global property was set - so we would need to think how that would work. We might either need to introduce a concept of targets that can't be run together in a design-time build, or manually run this target ourselves to gather references.

**Comment by srivatsn** [comment]
Ah it's a global property - understood.

**Comment by binki** [comment]
@srivatsn 
> We now get the packagereferences from a target called `CollectProjectReferences`. Overriding that target should be sufficient to restore both in the commandline and in VS.

Do you mean `CollectPackageReferences`? I am actually interested in `CollectProjectReferences` functionality and it’s confusing that you suggest such a thing exists when it doesn’t ;-).

**Comment by srivatsn** [comment]
Sorry yes that was a typo. Fixed inline.
